File: examples/PythonExamples/decorator.py
def catch_err(some_fn):
    def wrap():
        try:
            some_fn()
        except Exception as e:
            print(e)
    return wrap


@catch_err
def hi():
    print("Hello World")
    raise TypeError("类型报错")


hi()


"""
装饰器 @catch_err 等同于 catch_err(hi)()
"""

fn = catch_err(hi)
fn()

File: examples/typescript/basics-exam/src/decorator.ts
/*
  方法装饰器
 */
function catchErr() {
  return function (
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor,
  ): void {
    const original = descriptor.value;
    descriptor.value = function () {
      try {
        original.apply(this);
      } catch (e) {
        console.log(e);
      }
    };
  };
}

class SomeClass {
  @catchErr()
  hello() {
    console.log('Hello World!');
    throw new Error('发送错误');
  }
}

const someClass = new SomeClass();
someClass.hello();
